modules/mtl/arrow-mtl-data/src/test/kotlin/arrow/mtl/StateTTests.kt

```diff
@@ -1,64 +1,99 @@
+import arrow.core.extensions.listk.eqK.eqK
 import arrow.core.extensions.listk.monad.monad
+import arrow.core.extensions.listk.monadCombine.monadCombine
 import arrow.core.extensions.listk.semigroupK.semigroupK
+import arrow.core.extensions.option.eq.eq
+import arrow.core.extensions.tuple2.eq.eq
+import arrow.core.fix
 import arrow.fx.ForIO
 import arrow.fx.IO
-import arrow.fx.extensions.io.applicativeError.attempt
 import arrow.fx.extensions.io.async.async
 import arrow.fx.extensions.io.monad.monad
 import arrow.fx.mtl.statet.async.async
 import arrow.mtl.extensions.StateTMonadState
-import arrow.core.extensions.listk.monadCombine.monadCombine
 import arrow.mtl.extensions.statet.applicative.applicative
 import arrow.mtl.extensions.statet.monadCombine.monadCombine
 import arrow.mtl.extensions.statet.monadState.monadState
 import arrow.mtl.extensions.statet.semigroupK.semigroupK
 import arrow.test.UnitSpec
+import arrow.test.generators.GenK
+import arrow.test.generators.tuple2
 import arrow.test.laws.AsyncLaws
 import arrow.test.laws.MonadCombineLaws
 import arrow.test.laws.MonadStateLaws
 import arrow.test.laws.SemigroupKLaws
 import arrow.typeclasses.Eq
+import arrow.typeclasses.EqK
+import arrow.typeclasses.Monad
+import io.kotlintest.properties.Gen
 
 class StateTTests : UnitSpec() {
 
   val M: StateTMonadState<ForTry, Int> = StateT.monadState(Try.monad())
 
-  val EQ: Eq<StateTOf<ForTry, Int, Int>> = Eq { a, b ->
-    a.runM(Try.monad(), 1) == b.runM(Try.monad(), 1)
-  }
+  fun <F, S> eqk(EQKF: EqK<F>, EQS: Eq<S>, M: Monad<F>, s: S) = object : EqK<StateTPartialOf<F, S>> {
+    override fun <A> Kind<StateTPartialOf<F, S>, A>.eqK(other: Kind<StateTPartialOf<F, S>, A>, EQ: Eq<A>): Boolean =
+      (this.fix() to other.fix()).let {
+        val ls = it.first.runM(M, s)
+        val rs = it.second.runM(M, s)
 
-  val EQ_UNIT: Eq<StateTOf<ForTry, Int, Unit>> = Eq { a, b ->
-    a.runM(Try.monad(), 1) == b.runM(Try.monad(), 1)
+        EQKF.liftEq(Tuple2.eq(EQS, EQ)).run {
+          ls.eqv(rs)
+        }
+      }
   }
 
-  val EQ_LIST: Eq<Kind<StateTPartialOf<ForListK, Int>, Int>> = Eq { a, b ->
-    a.runM(ListK.monad(), 1) == b.runM(ListK.monad(), 1)
+  val listkStateEQK = eqk(ListK.eqK(), Int.eq(), ListK.monad(), 1)
+
+  val ioStateEQK = eqk(IO.eqK(), Int.eq(), IO.monad(), 1)
+
+  val tryEQK = object : EqK<ForTry> {
+    override fun <A> Kind<ForTry, A>.eqK(other: Kind<ForTry, A>, EQ: Eq<A>): Boolean =
+      (this.fix() to other.fix()).let {
+        Option.eq(EQ).run {
+          it.first.toOption().eqv(it.second.toOption())
+        }
+      }
   }
 
-  private fun <A> IOEQ(): Eq<StateTOf<ForIO, Int, A>> = Eq { a, b ->
-    a.runM(IO.monad(), 1).attempt().unsafeRunSync() == b.runM(IO.monad(), 1).attempt().unsafeRunSync()
+  val tryStateEqK: EqK<Kind<Kind<ForStateT, ForTry>, Int>> = eqk(tryEQK, Int.eq(), Try.monad(), 1)
+
+  fun <F, S> genk(genkF: GenK<F>, genS: Gen<S>) = object : GenK<StateTPartialOf<F, S>> {
```

Comment:
This is happening because we use `ListK` as the effect in `StateT`. The fully expanded and transformed type of `StateT<ForListK, Int, Int>` is: `Int -> ListK<Tuple2<Int, Int>>` And as soon as the lists get larger (by generation or by flatMapping etc) this test will run slower. Previously this only ran with lists of size one, which is awful for testdata-coverage, but is quite fast.

Imo using a different `F` for `StateT` instead of `ListK` is the best solution. You could also try to add a small list `GenK` which generates only lists < 10 size, but this will still grow quite large on stacksafety tests etc.

I think the best solution is to use `Option` instead of `List`. Tried it and the tests pass, also pretty straightforward to replace it

Reply:
ok, changed it.